File: Exercises2024/Chapter04_basic_mc/C/3-HardDisks/static/Source/sample.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "system.h"

#define MAXIMUM_NUMBER_OF_BINS 500

// samples the radial distribution function
void Sample(int Ichoise)
{
  int i,j;
  static double Distribution[MAXIMUM_NUMBER_OF_BINS];
  static double Count,Delta=0.0;
  double r2;
  VECTOR dr;
  FILE *FilePtr;

  switch(Ichoise)
  {
    case INITIALIZE:
      for(i=0;i<MAXIMUM_NUMBER_OF_BINS;i++)
        Distribution[i]=0.0;

      Count=0.0;
      // Delta  = Binsize
      Delta=0.5*BOXSIZE/(MAXIMUM_NUMBER_OF_BINS-1.0);
      break;
    case SAMPLE:
      // Sample The Radial Distribution Function
      // Loop Over All Particle Pairs
      // See Frenkel/Smit P. 77

      // Start Modification

      // End   Modification
      break;
    case WRITE_RESULTS:
      // write results to disk
      FilePtr=fopen("rdf.dat","w");
      for(i=0;i<MAXIMUM_NUMBER_OF_BINS-1;i++)
      {
        r2=M_PI*SQR(Delta)*NumberOfParticles*(NumberOfParticles-1)*(SQR(i+1)-(SQR(i)));
        fprintf(FilePtr,"%f %f\n",(i+0.5)*Delta,Distribution[i]*SQR(BOXSIZE)/(r2*Count));
      }
      fclose(FilePtr);
  }
}

File: Exercises2024/Chapter04_basic_mc/Fortran/3-HardDisks/static/Source/sample.f
      Subroutine Sample(Ichoise)
      Implicit None

      Include 'system.inc'

Cccccccccccccccccccccccccccccccccccccccccccccccccc
C     Samples The Radial Distribution Function   C
Cccccccccccccccccccccccccccccccccccccccccccccccccc

      Integer I,J,Maxx,Ichoise,A
      Double Precision BoxSize

      Parameter(Maxx = 500)
      Parameter(BoxSize = 10.0)

      Double Precision Ngr,Gg(Maxx),Delta,R2,DX,DY,R,ab,nid

      Save Ngr,Gg,Delta

      If(Ichoise.Eq.1) Then
         Do I=1,Maxx
            Gg(I) = 0.0d0
         Enddo

         Ngr   = 0.0d0
         Delta = 5.0d0/Dble(Maxx-1)
      
      Elseif(Ichoise.Eq.2) Then

Ccccccccccccccccccccccccccccccccccccccccccccccccc
C     Sample The Radial Distribution Function   C
C     Loop Over All Particle Pairs              C
C     See Frenkel/Smit P. 77                    C
C                                               C
C     Delta  = Binsize                          C
C     Ggt/Gg = Counters                         C
C     Maxx   = Maximum Number Of Bins           C
Ccccccccccccccccccccccccccccccccccccccccccccccccc

C     Start Modification

               
C     End   Modification

      Else

Ccccccccccccccccccccccccccccccc
C     Write Results To Disk   C
Ccccccccccccccccccccccccccccccc

         Do I=1,Maxx-1

            r=Delta*(I+0.5d0)
            ab=((i+1)**2-i**2)*Delta**2
            nid= 4.0d0*Datan(1.0d0)*Dble(Npart-1)/(BoxSize*BoxSize)*ab
            Gg(I)=Gg(I)/(Ngr*Dble(Npart)*nid)
         
            Write(21,*) r,Gg(I)
         Enddo
      Endif

      Return
      End
